crates/ruma-events/src/call/member.rs

```diff
@@ -1,324 +1,240 @@
+    /// The legacy format for m.call.member events. (An array of memberships. The devices of one
+    /// user.)
+    LegacyContent(LegacyMembershipContent),
+    /// Normal membership events. One event per membership. Multiple state keys will
+    /// be used to describe multiple devices for one user.
+    SessionContent(SessionMembershipData),
+    /// An empty content means this user has been in a rtc session but is not anymore.
+    Empty(EmptyMembershipData),
 }
 
 impl CallMemberEventContent {
-    /// Creates a new `CallMemberEventContent`.
-    pub fn new(memberships: Vec<Membership>) -> Self {
-        Self { memberships }
+    /// Creates a new [`CallMemberEventContent`] with [`LegacyMembershipData`].
+    pub fn new_legacy(memberships: Vec<LegacyMembershipData>) -> Self {
+        Self::LegacyContent(LegacyMembershipContent {
+            memberships, //: memberships.into_iter().map(MembershipData::Legacy).collect(),
+        })
     }
 
+    /// Creates a new [`CallMemberEventContent`] with [`SessionMembershipData`].
+    pub fn new(
+        application: Application,
+        device_id: String,
+        focus_active: ActiveFocus,
+        foci_preferred: Vec<Focus>,
+        created_ts: Option<MilliSecondsSinceUnixEpoch>,
+    ) -> Self {
+        Self::SessionContent(SessionMembershipData {
+            application,
+            device_id,
+            focus_active,
+            foci_preferred,
+            created_ts,
+        })
+    }
+
+    /// Creates a new Empty [`CallMemberEventContent`] representing a left membership.
+    pub fn new_empty(leave_reason: Option<LeaveReason>) -> Self {
+        Self::Empty(EmptyMembershipData { leave_reason })
+    }
     /// All non expired memberships in this member event.
     ///
-    /// In most cases you want tu use this method instead of the public memberships field.
+    /// In most cases you want to use this method instead of the public memberships field.
     /// The memberships field will also include expired events.
     ///
+    /// This copies all the memberships and converts them
     /// # Arguments
     ///
-    /// * `origin_server_ts` - optionally the `origin_server_ts` can be passed as a fallback in case
-    ///   the Membership does not contain `created_ts`. (`origin_server_ts` will be ignored if
-    ///   `created_ts` is `Some`)
+    /// * `origin_server_ts` - optionally the `origin_server_ts` can be passed as a fallback in the
+    ///   Membership does not contain [`LegacyMembershipData::created_ts`]. (`origin_server_ts` will
+    ///   be ignored if [`LegacyMembershipData::created_ts`] is `Some`)
     pub fn active_memberships(
         &self,
         origin_server_ts: Option<MilliSecondsSinceUnixEpoch>,
-    ) -> Vec<&Membership> {
-        self.memberships.iter().filter(|m| !m.is_expired(origin_server_ts)).collect()
+    ) -> Vec<MembershipData<'_>> {
+        match self {
+            CallMemberEventContent::LegacyContent(content) => content
```

Comment:
I believe it makes sense to implement `active_memberships` on `LegacyMembershipContet` and call it here.